pkg/server/server.go

```diff
@@ -73,9 +77,33 @@ func (c *ProxyClientConnection) send(pkt *client.Packet) error {
 }
 
 func NewPendingDialManager() *PendingDialManager {
-	return &PendingDialManager{
+	pm := &PendingDialManager{
 		pendingDial: make(map[int64]*ProxyClientConnection),
 	}
+
+	// spawn a goroutine to verify connection waiting is timeout or not.
+	go func() {
+		ticker := time.NewTicker(3 * time.Second)
```

Comment:
I would guess that most of the time connections resolve quickly. The would suggest that if we walk the pendingDials (we are anyway) we could determine minimum interval before the next connection request would expire and it would be close to the ConnectionSetupTimeout. That would suggest we would be better waiting the max of 3  seconds or ConnectionSetupTimeout. This reduces how often we have to grab the lock and walk the list.

Reply:
Food for thought: I'm wondering if it'll make sense to add another data structure that could sort the PendingDials by their start time. Then we'd only need to iterate through the oldest n entries every time.

Reply:
Depends on the length of the list. If we find it gets long then yes it would make sense. My guess (and its only a guess) is that usually requests spend very little time in the pendingDials list. If that's true then the length should be short and it may be mostly wasted effort. If it turns out to be false, then I think it would be a good change.